### PR TITLE
Codechange: Use std::array/std::unique_ptr for Persistent/TemporaryStorageArray.

### PR DESCRIPTION
**Pull request by PeterN**
## Motivation / Problem

Reducing use of C-style memory allocation and handling, memcpy, memset, sizeof, etc.

## Description

Replace the arrays/pointers in Persistent/TemporaryStorageArray with std::arrays instead. This allows for default blank initialization, and use of assignment operators to copy/reset the data. The constructors and destructors are no longer necessary.

Additionally the method "ResetToZero()" was never called, so is removed.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

File: src/newgrf_storage.h [hunk 1]
```diff
@@ -65,26 +65,10 @@ struct BasePersistentStorageArray {
...
-		free(this->prev_storage);
-	}
-
-	/** Resets all values to zero. */
-	void ResetToZero()
-	{
-		memset(this->storage, 0, sizeof(this->storage));
-	}
+	StorageType storage{}; ///< Memory for the storage array
+	std::unique_ptr<StorageType> prev_storage{}; ///< Memory to store "old" states so we can revert them on the performance of test cases for commands etc.
```
Comment (TrueBrain) [inline src/newgrf_storage.h]: "on the performance of test cases" .. it took me a while to parse that sentence :D Maybe reword it a bit to read a bit more natural? "so we can revert after running commands in test-mode" or something? Dunno, but "performance" threw me off here.

File: src/newgrf_storage.h [hunk 2]
```diff
@@ -104,10 +88,9 @@ struct PersistentStorageArray : BasePersistentStorageArray {
 
 		/* We do not have made a backup; lets do so */
 		if (AreChangesPersistent()) {
-			assert(this->prev_storage == nullptr);
-		} else if (this->prev_storage == nullptr) {
-			this->prev_storage = MallocT<TYPE>(SIZE);
-			memcpy(this->prev_storage, this->storage, sizeof(this->storage));
+			assert(!this->prev_storage);
+		} else if (!this->prev_storage) {
+			this->prev_storage = std::make_unique<StorageType>(this->storage);
```
Comment (TrueBrain) [inline src/newgrf_storage.h:93]: Why exactly a unique-pointer? If I read this correct, you kinda use it as an `std::optional`? Ofc you can use an unique-ptr as an optional, but initial I was thinking you actually need the pointer. But you just want to know if the object is set yes/no?

File: src/newgrf_storage.h (hunk 1 above)
Comment (PeterN) [reply src/newgrf_storage.h]: Not gonna lie, I never read all that comment as I didn't write it either...

File: src/newgrf_storage.h (hunk 2 above)
Comment (PeterN) [reply src/newgrf_storage.h:93]: As far as I know, using std::optional<> would (may?) mean prev_storage always takes up space even if it's not used. Given this 1KB each, using std::unique_ptr<> instead made sense to me.

You are right about not specifically needing the semantics of a pointer though.

**Review by TrueBrain — APPROVED** [review]
